Incremental commits off-chain

noonio opened the pull request:
:snowflake:   This PR implements _most_ of the off-chain changes we need for the incremental commits to work. Since there are many changes we decided to have a series of follow up PRs to finally conclude the wanted feature.


Included:

Relevant code to be able to do full end-to-end test for depositing (and recovering) a user UTxO into a running Head.
This means:
:snowflake:  everything needed to accept user requests at the API level
:snowflake:  build off-chain transactions (deposit, recover, increment)
:snowflake:   observe new transactions (deposit, recover and increment)
:snowflake:  changes in the `Snapshot` type and all other relevant types and functions to accommodate this new feature
:snowflake:  minimal needed changes in the on-chain code (**validators just return True so once this is merged we will not be able to release until this gap is implemented**)   

Outstanding:

:snowflake:  documentation - Write up some documentation to explain how incremental commits are working
:snowflake: Various tests described in the original [issue](https://github.com/cardano-scaling/hydra/issues/199) and which are relevant to the off-chain code
:snowflake: Make the increment/recover use the observed UTxO instead of keeping those around
:snowflake: Remove `Commit` client input since it is not used
:snowflake: Revisit the observation/transaction posting types and make sure the field names make sense
:snowflake: Determine what should be good enough deadline for a deposit transaction (should we make it configurable or rely on the contestation deadline)


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [ ] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

ch1bo commented:
@noonio When we continue work here, we should take a step back again given the new overall design of #199 with details from #1571. The overall off-chain story has not changed much, but is basically separated in two parts:
* Part 1: `hydra-node` API for deposit/cancel analogous to #1574
* Part 2: `hydra-node` detects deposits, triggers `ReqSn` with approval and `incrementTx` construction & observation (as before)

v0d1ch committed:
Incremental Commits off-chain changes

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -405,21 +434,23 @@ prop_observeAnyTx :: Property
-            Contest ContestObservation{headId} -> transition === Transition.Contest .&&. Just headId === expectedHeadId
-            Fanout FanoutObservation{headId} -> transition === Transition.Fanout .&&. Just headId === expectedHeadId
+      forAllShow genChainStateWithTx showTransition $ \(_, otherSt, _, _) -> do
+        -- genericCoverTable [transition] $ do
```
Must not disable this. Otherwise the coverage will not work.

v0d1ch committed:
Update the protocol.md

v0d1ch committed:
Make the recover work with just txIn

v0d1ch committed:
Recover observation

v0d1ch committed:
Make all tests green

ffakenz commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -52,7 +52,9 @@ data ClosedDatum = ClosedDatum
   -- ^ Spec: η. Digest of snapshotted UTxO
-  , deltaUTxOHash :: Hash
+  , alphaUTxOHash :: Hash
+  , -- TODO: ^ Spec: η?. Digest of UTxO still to be committed
```
alphaUTxOHash is not expected, we should be able to use 𝜂′Δ only.

ffakenz commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -115,6 +117,20 @@ data ContestRedeemer
+  , snapshotNumber :: SnapshotNumber
+  -- ^ Spec: s
+  , numberOfCommitOutputs :: Integer
+  -- ^ Spec: m
```
There is no `m` in the spec.

This should be a tx input: 𝜙deposit

ffakenz replied:
actually, at this moment, we can also removed this type completely.

ffakenz commented on hydra-plutus/src/Hydra/Contract/HeadError.hs:
```diff
@@ -50,6 +50,7 @@ data HeadError
   | LowerBoundBeforeContestationDeadline
   | FanoutNoLowerBoundDefined
   | FanoutUTxOToDecommitHashMismatch
+  | FanoutUTxOToCommitHashMismatch
```
this error seems not to be used.

otoh, we cannot tell during fannout if the hash corresponds to a decrement/increment.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Deposit.hs:
```diff
@@ -65,12 +65,13 @@ PlutusTx.unstableMakeIsData ''DepositDatum
 validator depositDatum r ctx =
   case r of
-    Claim -> False
+    Claim -> True
```
Could: mark this as `FIXME` so that we don't forget about it

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -221,6 +223,16 @@ commitDatum input = do
+  OpenDatum ->
+  IncrementRedeemer ->
+  Bool
+checkIncrement _ctx _openBefore _redeemer = True
```
Could: mark this as FIXME so that we don't forget abou

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -2153,6 +2391,8 @@ components:
           - $ref: "api.yaml#/components/schemas/UTxO"
           - $ref: "api.yaml#/components/schemas/FullCommitRequest"
+          - $ref: "api.yaml#/components/schemas/IncrementalCommitDepositRequest"
+          - $ref: "api.yaml#/components/schemas/IncrementalCommitRecoverRequest"
```
Must: not allow recover request body on `POST /commit`

ch1bo replied:
Also do not do so many indirections and I'm missing documentation for `DELETE /commits/<tx-in>`

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -2153,6 +2391,8 @@ components:
         oneof:
           - $ref: "api.yaml#/components/schemas/UTxO"
           - $ref: "api.yaml#/components/schemas/FullCommitRequest"
+          - $ref: "api.yaml#/components/schemas/IncrementalCommitDepositRequest"
```
Must: allow the same request body for deposit as for commit -> no need for a new schema

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -58,13 +70,27 @@ data DraftCommitTxRequest tx
+      }
+  | IncrementalCommitRecoverRequest
+      { recoverStart :: Natural
+      }
```
Must: not allow recover request body on the POST commit endpoint

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -163,42 +209,110 @@ httpApp tracer directChain pparams getInitializingHeadId getConfirmedUTxO putCli
+    Left err ->
+      pure (responseLBS status400 [] (Aeson.encode $ Aeson.String $ pack err))
+    Right someCommitRequest ->
+      getCommitInfo >>= \case
```
Must: not depend on the head state (the commit info projection)

We can always recover as long as the specified tx in is valid / known.

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -163,42 +209,110 @@ httpApp tracer directChain pparams getInitializingHeadId getConfirmedUTxO putCli
+        IncrementalCommit _ -> do
+          case someCommitRequest of
+            IncrementalCommitRecoverRequest{recoverStart} -> do
+              case checkRecover recoverQuery of
```
Must: use the `path`

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -163,42 +209,110 @@ httpApp tracer directChain pparams getInitializingHeadId getConfirmedUTxO putCli
+      Right txInStr ->
+        case Aeson.eitherDecode (fromStrict txInStr) :: Either String TxIn of
+          Left e -> Left (responseLBS status400 [] (Aeson.encode $ Aeson.String $ "Cannot recover funds. Failed to parse TxIn: " <> pack e))
+          Right txIn -> Right txIn
```
Note that there is a `parseTxIn` function. Although the `#` might be problematic in a URL path, so we should just require a `TxId`

ch1bo commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -163,42 +209,110 @@ httpApp tracer directChain pparams getInitializingHeadId getConfirmedUTxO putCli
+      getCommitInfo >>= \case
+        IncrementalCommit _ -> do
+          case someCommitRequest of
+            IncrementalCommitRecoverRequest{recoverStart} -> do
```
Must: not require a starting slot number from the user

Just use the `currenSlot` from the time handle (i.e. in the chain handler or so)

v0d1ch committed:
Remove IncrementalCommitRecoverRequest

We don't need to accept lower boundary slot number also

v0d1ch committed:
Init incremental commit demo instructions

ch1bo committed:
Complete instructions for a basic incremental commit

v0d1ch committed:
Add export

v0d1ch committed:
Remove IncrementalCommitDepositRequest

v0d1ch committed:
Use path instead of query string for recover

ch1bo committed:
Add instructions to recover a deposit

ch1bo committed:
Define genRecoverTx in terms of genDepositTx

This also enables re-enables observation properties of Deposit/Recover
transitions. We note though, that these are not "real" head state
transitions and the ctx / state return values are superficial.

ch1bo committed:
Change definition of recoverTx and move observeDepositTx to hydra-tx

This makes it easier to use in combination with observing a deposit.

v0d1ch committed:
Move recover observation to hydra-tx

Remove deposit and recover from the Head tx observation tests and have
dedicated tests for deposit/recover.

v0d1ch committed:
Resolve fixme in deposit observation

Make sure to check if the deposit datum actually contains the deposit
tx inputs.
Remove the `utxo` from `DepositObservation` and propagate this change.
Increment observation is broken now since it was depending on this utxo
but we need to do it differently any way.

v0d1ch committed:
Keep Map TxIn locally so we support multiple deposits

Added TxInType since we needed abstract type for use in head logic.
Ideally I would not try to find utxo in the map but deal with txin/id
only on increment/recover

v0d1ch committed:
Add deposit TxIn to Increment and OnIncrementTx

If we have this type around it will serve us well for deleting
observed deposits and also on recovers.

v0d1ch committed:
Add extra UTxO to make the observations pass

On increment the UTxO we are incrementing is not part of the head UTxO
yet so we need more information in order to be able to observe the
relevant transaction.

v0d1ch committed:
Make recover workflow go through the HeadLogic

v0d1ch committed:
Remove alphaUTxOHash datum field and fields in IncrementRedeemer

v0d1ch committed:
Remove unused error constructor

v0d1ch committed:
Add a test for pending deposits

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -721,6 +721,74 @@ canRecoverDeposit tracer workDir node hydraScriptsTxId =
+              >>= httpJSON
+
+          -- the issue doesn't specify the format of the response so we are
+          -- free to use whatever is convenient for the users ([TxIn]?)
```
It should respond with anything sufficient to call `DELETE /commits/<txid>` endpoint. On top of that any additional information about the commit (e.g. until when its locked) would be beneficial.

v0d1ch committed:
Add a test that spends the script into a head

noonio committed:
Implement GET endpoint for pending deposits

v0d1ch committed:
Make everything compile

v0d1ch committed:
Add commit blueprint for deposit workflow

v0d1ch committed:
Use TxId instead of TxIn everywhere (recover/deposit related)

v0d1ch committed:
Fix tests relying on empty utxo hash

v0d1ch committed:
Make all tests green

v0d1ch committed:
Finish up the tutorial

v0d1ch committed:
Delete from pending deposits on successfull recover

v0d1ch committed:
Fix a bug in the pending deposit projection

v0d1ch committed:
Fix json schema errors

v0d1ch committed:
Make protocol.md true to implementation

v0d1ch committed:
Fix plutus scripts

locallycompact commented on hydra-node/src/Hydra/API/HTTPServer.hs:
```diff
@@ -84,7 +92,11 @@ instance (FromJSON tx, FromJSON (UTxOType tx)) => FromJSON (DraftCommitTxRequest
 
 instance (Arbitrary tx, Arbitrary (UTxOType tx)) => Arbitrary (DraftCommitTxRequest tx) where
-  arbitrary = genericArbitrary
+  arbitrary =
```
TODO: revert

v0d1ch committed:
Wrap things up

ch1bo commented on CHANGELOG.md:
```diff
@@ -10,6 +10,17 @@ changes.
+      - Add documentation to explain the feature
+      - Implement tests scenarios outlined in the [incremental-commit] (https://github.com/cardano-scaling/hydra/issues/199) issue
+      - Remove Commit client input since it is unused
+      - Revisit types related to observations/posting transactions and make sure the fields are named appropriatelly
```
:heart:
